[FB] [PI-3478] Metadata and cube maths documentation (#3869)
SciTools/iris: docs/iris/src/further_topics/metadata.rst

Iris models several other classes of `CF Conventions`_ metadata apart from
those mentioned above. Such as the :class:`~iris.coords.AncillaryVariable`
(see `Ancillary Data`_ and `Flags`_), the :class:`~iris.coords.CellMeasure`
(see `Cell Measures`_), and also the :class:`~iris.aux_factory.AuxCoordFactory`
(see `Parametric Vertical Coordinate`_).

Collectively, the asforementioned classes will be known here as the Iris
`CF Conventions` classes.

.. hint::

    If there are any `CF Conventions`_ metadata missing from Iris that you
    care about, then please let us know by raising a `GitHub Issue`_ on
    `SciTools/iris`_

Common Metadata
===============

What each of these **different** Iris `CF Conventions`_ classes all have in
**common** is that **metadata** is used to define them and give them meaning.

.. _metadata members:
.. table:: - Iris classes that model `CF Conventions`_ metadata
   :widths: auto
   :align: center

   =================== ======================================= ============================== ========================================== ================================= ======================== ============================== ===================

>    Metadata members    :class:`~iris.coords.AncillaryVariable` :class:`~iris.coords.AuxCoord` :class:`~iris.aux_factory.AuxCoordFactory` :class:`~iris.coords.CellMeasure` :class:`~iris.cube.Cube` :class:`~iris.coords.DimCoord` Metadata members

Author: The table is quite wide, and I can't do much about that... this is particularly a problem on a mobile device, so scrolling to the right it is easy to lose track of what rows are for what metadata members.

Repeating the column at the end of the table help alleviate this problem IMHO

   ``circular``                                                                                                                                                                                     ✔                              ``circular``
   =================== ======================================= ============================== ========================================== ================================= ======================== ============================== ===================

:numref:`metadata members` shows for each Iris `CF Conventions`_ class the
collective of individual metadata members used to define it. Almost all
of these members reference specific `CF Conventions`_ terms. However, some
of these members, such as :attr:`~iris.coords.DimCoord.var_name` and
:attr:`~iris.coords.DimCoord.circular` are Iris specific terms.

For example, the collective metadata used to define an
:class:`~iris.coords.AncillaryVariable` are the ``standard_name``, ``long_name``,
``var_name``, ``units``, and ``attributes`` members. Note that, these are the
actual `data attribute`_ names of the actual metadata members on the actual

> Iris class.

Author: It was intended to stress the fact, but if it reads clunkily, then happy to drop it

As :numref:`metadata members` highlights, **specific** metadata is used to
define and represent each **specific** Iris `CF Conventions`_ class. This means
that this **specific** metadata can then be used to easily **identify**,
**compare** and **differentiate** between individual class instances.

Common Metadata API
===================

.. testsetup::

    import iris
    cube = iris.load_cube(iris.sample_data_path("A1B_north_america.nc"))

As of Iris ``3.0.0``, a unified treatment of metadata has been applied

> across each Iris class in :numref:`metadata members` to allow users to easily

Author: Took a slightly difference route to clarifying this, see what you think...

    DimCoordMetadata(standard_name=None, long_name=None, var_name='longitude', units=None, attributes={'grinning face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Another very useful method from the `namedtuple`_ toolkit is `namedtuple._fields`_.
This method returns a tuple of strings listing the ``metadata`` members, in a
fixed order. This allows you to easily iterate over the metadata class members,
for what ever purpose you may require, e.g.,

    >>> metadata._fields
    ('standard_name', 'long_name', 'var_name', 'units', 'attributes', 'coord_system', 'climatological', 'circular')

    >>> tuple([getattr(metadata, member) for member in metadata._fields])
    ('longitude', None, 'longitude', Unit('degrees'), {'grinning face': '🙃'}, GeogCS(6371229.0), False, False)

    >>> tuple([getattr(metadata, member) for member in metadata._fields if member.endswith("name")])
    ('longitude', None, 'longitude')

Note that, `namedtuple._fields`_ is also a class method, so you don't need
an instance to determine the members of a metadata class, e.g.,

    >>> from iris.common import CubeMetadata
    >>> CubeMetadata._fields

>     ('standard_name', 'long_name', 'var_name', 'units', 'attributes', 'cell_methods')

Author: Fair point, but I think a lot of users haven't been exposed to namedtuples.

Also, if they have, then this highlights that you can easily rip apart the metadata, and iterate over it easily, rather than create your own enumerated list of members.

So, I think it has value... but that's based on speculation - that's why docs are hard as you don't know the experience level of the readership. One persons dross is another persons gold.

above and beyond that of the  standard Python `namedtuple`_, which allows you to
easily **compare**, **combine**, **convert** and understand the **difference**
between your ``metadata`` instances.

.. _metadata equality:

Metadata equality
^^^^^^^^^^^^^^^^^

The metadata classes support both **equality** (``__eq__``) and **inequality**
(``__ne__``), but no other `rich comparison`_ operators are implemented.
This is simply because there is no obvious ordering to any collective of metadata
members, as defined in :numref:`metadata members`.

For example, given the following :class:`~iris.coords.DimCoord`,

.. doctest:: richer-metadata

    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

We can compare ``metadata`` using the ``==`` operator, as you may naturally
expect,

.. doctest:: richer-metadata

    >>> longitude.metadata == longitude.metadata
    True

Or alternatively, using the ``equal`` method instead,

.. doctest:: richer-metadata

    >>> longitude.metadata.equal(longitude.metadata)

>     True

Author: I wanted to add a hook after highlighting `equal` as an alternative to `==`, as the `equal` methods supports the `lenient` kwarg, where as that's impossible for `__eq__`... I've not added that, as I'm still writing the section on lenient metadata.

As a reminder I'll add a `todo`, which we can merge into the feature branch, but I'll ensure that all `todo`s are removed before merging to `master`.

   =================== ======================================= ============================== ========================================== ================================= ======================== ============================== ===================

:numref:`metadata members table` shows for each Iris `CF Conventions`_ class the
collective of individual metadata members used to define it. Almost all
of these members reference specific `CF Conventions`_ terms. However, some
of these members, such as :attr:`~iris.coords.DimCoord.var_name` and
:attr:`~iris.coords.DimCoord.circular` are Iris specific terms.

For example, the collective metadata used to define an
:class:`~iris.coords.AncillaryVariable` are the ``standard_name``, ``long_name``,
``var_name``, ``units``, and ``attributes`` members. Note that, these are the
actual `data attribute`_ names of the actual metadata members on the actual
Iris class.

As :numref:`metadata members table` highlights, **specific** metadata is used to
define and represent each **specific** Iris `CF Conventions`_ class. This means
that this **specific** metadata can then be used to easily **identify**,

> **compare** and **differentiate** between individual class instances.

Author: Yeah, that is a bit `specific` tastic :rofl:

    True

Strict equality
"""""""""""""""

By default, metadata class equality will perform a **strict** comparison between
each associated ``metadata`` member. If **any** ``metadata`` member has a
different value, then the result of the operation will be ``False``. For example,

.. doctest:: richer-metadata

    >>> other = longitude.metadata._replace(standard_name=None)
    >>> other
    DimCoordMetadata(standard_name=None, long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
    >>> longitude.metadata == other
    False

.. doctest:: richer-metadata

    >>> longitude.attributes = {"grinning face": "🙂"}
    >>> other = longitude.metadata._replace(attributes={"grinning face":  "🙃"})
    >>> other
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
    >>> longitude.metadata == other
    False

Note that, although the ``==`` operator and the ``equal`` method are
functionally equivalent, the ``equal`` method also provides a means

> to support **lenient** equality, as discussed in :ref:`lenient metadata`.

Author: Yup, I'll move this back to the previous section, which will kill two birds with one stone, thanks.

.. doctest:: richer-metadata

    >>> metadata1 == metadata2
    True

.. doctest:: richer-metadata

    >>> metadata2 = cube.metadata._replace(attributes={"one": np.int(1), "two": np.array([0.1, 0.2])})
    >>> metadata1 == metadata2
    False

.. _compare like:

Comparing like with like
""""""""""""""""""""""""

So far in our journey through metadata class equality, we have only considered
cases where the operands are instances of the **same** type. It is possible to
compare instances of **different** metadata classes, but the result will always
be ``False``,

.. doctest:: richer-metadata

    >>> cube.metadata == longitude.metadata
    False

The reason different metadata classes cannot be compared is simply because each
metadata class contains **different** members, as shown in
:numref:`metadata members table`. However, there is an exception to the rule...

.. _exception rule:

Exception to the rule
~~~~~~~~~~~~~~~~~~~~~

In general, **different** metadata classes cannot be compared, however support

> is provided for comparing :class:`~iris.common.CoordMetadata` and

Author: Table 1 and 2 explain what `CoordMetadata` is and where exactly it appears in Iris.

The user will end up using `CoordMetadata` at some point if they ever use the common metadata API.

First, confirm that the ``metadata`` is different,

.. doctest:: richer-metadata

    >>> longitude.metadata != metadata
    True

As expected, the ``metadata`` is different. Now, let's answer the question,
"**what's the difference?**",

.. doctest:: richer-metadata

    >>> longitude.metadata.difference(metadata)
    DimCoordMetadata(standard_name=None, long_name=(None, 'lon'), var_name=('longitude', 'lon'), units=(Unit('degrees'), Unit('radians')), attributes=None, coord_system=None, climatological=None, circular=None)

The ``difference`` method returns a :class:`~iris.common.DimCoordMetadata` instance, when
there is **at least** one ``metadata`` member with a different value, where,

- ``None`` means that there was **no** difference for the member,
- a `tuple`_ containing the two different associated values for the member.

Given our example, only the ``long_name``, ``var_name`` and ``units`` members
have different values, as expected. Note that, the ``difference`` method **is

> not** commutative. The the order of the tuple member values is the same order

Author: Great spot :eyes:

with another instance that's identical apart from its ``standard_name`` member,
which is replaced with **different value**,

.. doctest:: metadata-combine

    >>> metadata = cube.metadata._replace(standard_name="air_pressure_at_sea_level")
    >>> metadata != cube.metadata
    True
    >>> metadata.combine(cube.metadata)  # doctest: +SKIP
    CubeMetadata(standard_name=None, long_name=None, var_name='air_temperature', units=Unit('K'), attributes={'STASH': STASH(model=1, section=3, item=236), 'source': 'Data from Met Office Unified Model 6.05', 'Model scenario': 'A1B', 'Conventions': 'CF-1.5'}, cell_methods=(CellMethod(method='mean', coord_names=('time',), intervals=('6 hour',), comments=()),))

The ``combine`` method combines metadata by performing a **strict** comparison
between each of the associated metadata member values,

- if the values are **different**, then the combined result is ``None``
- otherwise, the combined result is the **common value**

Let's reinforce this behaviour, but this time by combining metadata where the
``attributes`` `dict`_ member is different, where,

- the ``STASH`` and ``source`` keys are **missing**,

> - the ``Model scenario`` key has the same **same value**,

Author: Nice one :smile:

.. _metadata:

********
Metadata
********

This section provides a detailed overview of how your metadata is managed
within Iris. In particular, it discusses what metadata is, where it fits
into Iris, and more importantly how you can create, access, manipulate,
and analyse your metadata.

All the finer details covered here may not be entirely relevant to your use
case, but it's here if you ever need it. In fact, you may want to skip
straight ahead to :ref:`richer metadata`, and take it from there.

> Introduction

Reviewer: As I mentioned in my main comment:

It took me a while to follow the introduction and first section as I think the paragraphs hop around slightly.  All the vital information in there, so I doubt anyone else will notice, but I would:
* Introduction:  Move the first paragraph down, so all the class information is together, and give a more explicit reference to where the Cube, DimCoord, AuxCoord info is introduced wrt CF by re-referencing the "Data structures doc"

```suggestion
Introduction
============

As discussed in :ref:`iris_data_structures`, Iris draws heavily from the
`NetCDF CF Metadata Conventions`_ as a source for its data model, thus building
on the widely recognised and understood terminology defined within those
`CF Conventions`_ by the scientific community.

In :ref:`iris_data_structures` we also introduced several classes in Iris that care about your ``data``,
and also your ``metadata``: the :class:`~iris.cube.Cube`, the :class:`~iris.coords.AuxCoord`, and the
:class:`~iris.coords.DimCoord`.  In addition to these Iris models several other classes of `CF Conventions`_ metadata: the :class:`~iris.coords.AncillaryVariable`
(see `Ancillary Data`_ and `Flags`_), the :class:`~iris.coords.CellMeasure`
(see `Cell Measures`_), and also the :class:`~iris.aux_factory.AuxCoordFactory`
(see `Parametric Vertical Coordinate`_).

```

============

So far we've introduced several classes in Iris that care about your ``data``,
and also your ``metadata`` i.e., `data about data`_. Primarily, these classes
are the :class:`~iris.cube.Cube`, the :class:`~iris.coords.AuxCoord`, and the
:class:`~iris.coords.DimCoord`, all of which should be familiar to you now.

As discussed in :ref:`iris_data_structures`, Iris draws heavily from the
`NetCDF CF Metadata Conventions`_ as a source for its data model, thus building
on the widely recognised and understood terminology defined within those
`CF Conventions`_ by the scientific community.

Iris models several other classes of `CF Conventions`_ metadata apart from
those mentioned above. Such as the :class:`~iris.coords.AncillaryVariable`
(see `Ancillary Data`_ and `Flags`_), the :class:`~iris.coords.CellMeasure`
(see `Cell Measures`_), and also the :class:`~iris.aux_factory.AuxCoordFactory`
(see `Parametric Vertical Coordinate`_).

> Collectively, the asforementioned classes will be known here as the Iris

Reviewer: typo
```suggestion
Collectively, the aforementioned classes will be known here as the Iris
```

`CF Conventions` classes.

.. hint::

    If there are any `CF Conventions`_ metadata missing from Iris that you
    care about, then please let us know by raising a `GitHub Issue`_ on
    `SciTools/iris`_

> Common Metadata

Reviewer: As I mentioned in my main comment:

All the info is here, but I found it easier to understand the table once I'd understood what is meant by metadata.  So I'd introduce what a metadata is first, before one tries to read it.  So, I’d move the table to the end of the section, and the example describing it immediately before.
Also:
* its ambiguous as to whether "var_name" and "circular" are the only Iris specific terms.
* In the table we have "standard name" is this the same as the "standard_name" in the pros ? 
* Why do some of the metadata members listed in the table have underscores in them, but others don't.  Are they the same things?  Is one a method?
* And - what does the red represent - I've been wondering this throughout our documentation.

```suggestion
Common Metadata
===============

Each of the Iris `CF Conventions`_ classes use  **metadata** to define them and give them meaning.

A small collective of individual **metadata members** can be used to define the
Iris `CF Conventions`_ classes.  All the members reference specific `CF Conventions`_ terms, 
except for  :attr:`~iris.coords.DimCoord.var_name` and :attr:`~iris.coords.DimCoord.circular` which are Iris specific terms.  The individual metadata members used to define the various Iris `CF Conventions`_ classes are shown in :numref:`metadata members table` . 

As :numref:`metadata members table` highlights, **specific** metadata is used to
define and represent each of the Iris `CF Conventions`_ classes. Thus the
metadata can be used to easily **identify**, **compare** and **differentiate**
between individual class instances.

For example, the collective metadata used to define an
:class:`~iris.coords.AncillaryVariable` are the ``standard_name``, ``long_name``,
``var_name``, ``units``, and ``attributes`` members. Note that, these are the
actual `data attribute`_ names of the metadata members on the Iris class.

```
.. and now have Table 1.  Should the member names all have "_"'s?

    Traceback (most recent call last):
    AttributeError: can't set attribute

Note that, the ``metadata`` property re-creates a **new** instance per invocation,
with a **snapshot** of the container class metadata values at that point in time,

    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=True)

Metadata class behaviour
------------------------

As mentioned previously, the metadata classes in :numref:`metadata classes table`
inherit the behaviour of a `namedtuple`_, and so act and feel like a `namedtuple`_,
just as you might expect. For example, given the following ``metadata``,

    >>> metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

We can use the `namedtuple._make`_ method to create a **new**
:class:`~iris.common.metadata.DimCoordMetadata` instance from an existing sequence
or iterable:

>     >>> values = (1, 2, 3, 4, 5, 6, 7, 8)

Reviewer: Q: How do I know which metadata member the values are assigned to?  I'm assuming you have a secret hidden order?

Author: @abooton It's no secret, it's implied that it matches the `namedtuple._fields`. I've added clarification.

"""""""""""""""

By default, metadata class equality will perform a **strict** comparison between
each associated ``metadata`` member. If **any** ``metadata`` member has a
different value, then the result of the operation will be ``False``. For example,

.. doctest:: richer-metadata

    >>> other = longitude.metadata._replace(standard_name=None)
    >>> other
    DimCoordMetadata(standard_name=None, long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
    >>> longitude.metadata == other
    False

.. doctest:: richer-metadata

    >>> longitude.attributes = {"grinning face": "🙂"}
    >>> other = longitude.metadata._replace(attributes={"grinning face":  "🙃"})
    >>> other
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
    >>> longitude.metadata == other
    False

One further point worth highlighting is it is possible for `NumPy`_ scalars
and arrays to appear in the ``attributes`` `dict`_ of some Iris metadata class

> instances. Normally, this would cause issues,

Reviewer: ```suggestion
instances. Normally, this would cause issues. For example,
```

    Alternatively, use the ``from_metadata`` class method instead, see
    :ref:`metadata conversion`.

Comparing the instances confirms that equality is indeed supported between
:class:`~iris.common.metadata.DimCoordMetadata` and :class:`~iris.common.metadata.CoordMetadata`
classes,

.. doctest:: richer-metadata

    >>> latitude.metadata == metadata
    True

The reason for this behaviour is primarily historical. The ``circular``
member has **never** been used by the ``__eq__`` operator when comparing an
:class:`~iris.coords.AuxCoord` and a :class:`~iris.coords.DimCoord`. Therefore
for consistency, this behaviour is also extended to ``__eq__`` for the associated
container metadata classes.

However, note that the ``circular`` member **is used** by the ``__eq__`` operator
when comparing one :class:`~iris.coords.DimCoord` to another. This also applies
when comparing :class:`~iris.common.metadata.DimCoordMetadata`.

This exception to the rule for :ref:`equality <metadata equality>` also applies
to the :ref:`difference <metadata difference>` and :ref:`combine <metadata combine>`
methods of metadata classes.

.. _metadata difference:

Metadata difference
^^^^^^^^^^^^^^^^^^^

> Being able to compare metadata is valuable. Particularly when we have the

Reviewer: ```suggestion
Being able to compare metadata is valuable, especially when we have the
```

convenience of being able to do this easily with metadata classes. However,
when the result of comparing two metadata instances is ``False``, it begs

> the next obvious question, "**what's the difference?**"

Reviewer: ```suggestion
the question, "**what's the difference?**"
```

different values,

.. doctest:: richer-metadata

    >>> from cf_units import Unit
    >>> metadata = longitude.metadata._replace(long_name="lon", var_name="lon", units=Unit("radians"))
    >>> metadata
    DimCoordMetadata(standard_name='longitude', long_name='lon', var_name='lon', units=Unit('radians'), attributes={'grinning face': '🙂'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

First, confirm that the ``metadata`` is different,

.. doctest:: richer-metadata

    >>> longitude.metadata != metadata
    True

As expected, the ``metadata`` is different. Now, let's answer the question,
"**what's the difference?**",

.. doctest:: richer-metadata

    >>> longitude.metadata.difference(metadata)
    DimCoordMetadata(standard_name=None, long_name=(None, 'lon'), var_name=('longitude', 'lon'), units=(Unit('degrees'), Unit('radians')), attributes=None, coord_system=None, climatological=None, circular=None)

The ``difference`` method returns a :class:`~iris.common.metadata.DimCoordMetadata` instance, when
there is **at least** one ``metadata`` member with a different value, where,

- ``None`` means that there was **no** difference for the member,

> - a `tuple`_ containing the two different associated values for the member

Reviewer: ```suggestion
- a `tuple`_ contains the two different associated values for the member
```

.. doctest:: richer-metadata

    >>> attributes = {"grinning face": "😀", "neutral face": "😐"}
    >>> longitude.attributes = attributes
    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '😀', 'neutral face': '😐'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Then create another :class:`~iris.common.metadata.DimCoordMetadata` with a different
``attributes`` `dict`_, namely,

- the ``grinning face`` key has the **same value**,
- the ``neutral face`` key has a **different value**, and
- the ``upside-down face`` key is **new**

.. doctest:: richer-metadata

    >>> attributes = {"grinning face": "😀", "neutral face": "😜", "upside-down face": "🙃"}
    >>> metadata = longitude.metadata._replace(attributes=attributes)
    >>> metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={'grinning face': '😀', 'neutral face': '😜', 'upside-down face': '🙃'}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

> Now, let's compare the two above instances for differences, and see what we get,

Reviewer: I was being slow...

```suggestion
Now, let's compare the two above instances for differences, and we see the attribute member contains a tuple of the differing information as expected, but each value is a dictionary containing the differing key/values,
```

Author: @abooton I wasn't as verbose, as I don't quite think was quite necessary, but I've updated to highlight that the difference is in the `attributes` dictionary. Hope that's a happy compromise...

^^^^^^^^^^^^^^^^^^^^

.. testsetup:: metadata-combine

   import iris
   cube = iris.load_cube(iris.sample_data_path("A1B_north_america.nc"))
   longitude = cube.coord("longitude")

So far we've seen how to :ref:`compare metadata <metadata equality>`, and also how
to determine the :ref:`difference between metadata <metadata difference>`. Now we
take the next step, and explore how to combine metadata together using the ``combine``
metadata class method.

For example, consider the following :class:`~iris.common.metadata.CubeMetadata`,

.. doctest:: metadata-combine

    >>> cube.metadata  # doctest: +SKIP
    CubeMetadata(standard_name='air_temperature', long_name=None, var_name='air_temperature', units=Unit('K'), attributes={'Conventions': 'CF-1.5', 'STASH': STASH(model=1, section=3, item=236), 'Model scenario': 'A1B', 'source': 'Data from Met Office Unified Model 6.05'}, cell_methods=(CellMethod(method='mean', coord_names=('time',), intervals=('6 hour',), comments=()),))

We can perform the **identity function** by comparing the metadata with itself,

.. doctest:: metadata-combine

    >>> metadata = cube.metadata.combine(cube.metadata)
    >>> cube.metadata == metadata
    True

> As you might suspect, combining identical metadata returns metadata that is

Reviewer: Q:  Do you mean "As you might expect" rather that "as expected" because, one might assume the metadata gets smashed together?  (As in leinient data? - I don't know as I've not read that section yet).

:class:`~iris.coords.AuxCoord` and :class:`~iris.coords.DimCoord`,

.. doctest:: metadata-combine

    >>> forecast_period = cube.coord("forecast_period")
    >>> longitude = cube.coord("longitude")

First, let's see their associated metadata,

.. doctest:: metadata-combine

    >>> forecast_period.metadata
    CoordMetadata(standard_name='forecast_period', long_name=None, var_name='forecast_period', units=Unit('hours'), attributes={}, coord_system=None, climatological=False)
    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Before combining their metadata together,

.. doctest:: metadata-combine

    >>> forecast_period.metadata.combine(longitude.metadata)
    CoordMetadata(standard_name=None, long_name=None, var_name=None, units=None, attributes={}, coord_system=None, climatological=False)
    >>> longitude.metadata.combine(forecast_period.metadata)
    DimCoordMetadata(standard_name=None, long_name=None, var_name=None, units=None, attributes={}, coord_system=None, climatological=False, circular=None)

> However, note that commutativity in this case cannot be honoured, for obvious reasons.

Reviewer: Haha - the obvious reason is because the time can't have circular?  How have you managed to make it intelligently understand this!!  That's pretty cool given it isn't to do with the type of coord, but the scientific nature of the data being processed.

.. _metadata conversion:

Metadata conversion
^^^^^^^^^^^^^^^^^^^

.. testsetup:: metadata-convert

   import iris
   from iris.common import DimCoordMetadata
   cube = iris.load_cube(iris.sample_data_path("A1B_north_america.nc"))
   longitude = cube.coord("longitude")

In general, the :ref:`equal <metadata equality>`, :ref:`difference <metadata difference>`,
and :ref:`combine <metadata combine>` methods only support operations on instances
of the same metadata class (see :ref:`exception to the rule <exception rule>`).

> However, metadata may be converted from one metadata class to another by using

Reviewer: ```suggestion
However, metadata may be converted from one metadata class to another using
```

:class:`~iris.common.metadata.CubeMetadata` class does not contain the following
:class:`~iris.common.metadata.DimCoordMetadata` members,

- ``coords_system``
- ``climatological``, and
- ``circular``

Thus these particular metadata members are set to ``None`` in the resultant
:class:`~iris.common.metadata.DimCoordMetadata` instance.

Note that, the ``from_metadata`` method is also available on a metadata
class instance,

.. doctest:: metadata-convert

    >>> longitude.metadata.from_metadata(cube.metadata)
    DimCoordMetadata(standard_name='air_temperature', long_name=None, var_name='air_temperature', units=Unit('K'), attributes={'Conventions': 'CF-1.5', 'STASH': STASH(model=1, section=3, item=236), 'Model scenario': 'A1B', 'source': 'Data from Met Office Unified Model 6.05'}, coord_system=None, climatological=None, circular=None)

Metadata assignment
^^^^^^^^^^^^^^^^^^^

.. testsetup:: metadata-assign

   import iris
   cube = iris.load_cube(iris.sample_data_path("A1B_north_america.nc"))
   longitude = cube.coord("longitude")
   original = longitude.copy()
   latitude = cube.coord("latitude")

The ``metadata`` property available on each Iris `CF Conventions`_ container

> class (see :numref:`metadata classes table`) can be use not only **to get**

Reviewer: ```suggestion
class (see :numref:`metadata classes table`) can not only be used **to get**
```

the metadata of an instance, but also **to set** the metadata on an instance.

> 

Reviewer: For some reason I did not pickup on the difference between longitude and latitude.  Maybe just one line and one box would be ok:

```
For example, we can assign the :class:`~iris.common.metadata.DimCoordMetadata` metadata from the
``longitude`` coordinate directly to the ``latitude`` coordinate,

.. doctest:: metadata-assign

    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
    >>> longitude.metadata = latitude.metadata
    >>> longitude.metadata
    DimCoordMetadata(standard_name='latitude', long_name=None, var_name='latitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)
```

Author: @abooton I've printed the `latitude.metadata` just to clarigy, what's what...

For example, given the following :class:`~iris.common.metadata.DimCoordMetadata` of the
``longitude`` coordinate,

.. doctest:: metadata-assign

    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

We can assign to it directly using the :class:`~iris.common.metadata.DimCoordMetadata` of the ``latitude``
coordinate,

.. doctest:: metadata-assign

    >>> longitude.metadata = latitude.metadata
    >>> longitude.metadata

>     DimCoordMetadata(standard_name='latitude', long_name=None, var_name='latitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Reviewer: Q:  What happens to the metadata-data values, are we trying to put lat values on longs?
Lats -90 to 90.  Lons go -180 to 180.  Do you end up with a 90 by 90 deg region?

Author: @abooton This is just an example of assigning metadata... it's really a bit of a nonsense, but conveniently serves the purpose of highlighting the utility of being easily able to assign **all** the metadata from one instance to another... regardless of whether it makes semantic sense or not i.e., yup, indeed, you now have a `longitude` points/bounds for a `DimCoord` that according to the metadata is a `latitude`

classes share the following common metadata members,

- ``standard_name``,
- ``long_name``,
- ``var_name``,
- ``units``,
- ``attributes``

As such, all of these metadata members of the resultant
:class:`~iris.common.metadata.DimCoordMetadata` instance are populated from the associated
:class:`~iris.common.metadata.CubeMetadata` instance members. However, a
:class:`~iris.common.metadata.CubeMetadata` class does not contain the following
:class:`~iris.common.metadata.DimCoordMetadata` members,

- ``coords_system``,
- ``climatological``,
- ``circular``

Thus these particular metadata members are set to ``None`` in the resultant
:class:`~iris.common.metadata.DimCoordMetadata` instance.

Note that, the ``from_metadata`` method is also available on a metadata
class instance,

.. doctest:: metadata-convert

    >>> longitude.metadata.from_metadata(cube.metadata)
    DimCoordMetadata(standard_name='air_temperature', long_name=None, var_name='air_temperature', units=Unit('K'), attributes={'Conventions': 'CF-1.5', 'STASH': STASH(model=1, section=3, item=236), 'Model scenario': 'A1B', 'source': 'Data from Met Office Unified Model 6.05'}, coord_system=None, climatological=None, circular=None)

> 

Author: @rcomer Yup, it is, but only once its assigned to the class container... subtle difference, and I didn't want to get into that in the docs. Interesting that you picked up on it :100: 

Just for your own weirded out sanity... that's the job of the `iris.common.mixin.CFVariableMixin` which will **validate** any content assigned to the class container (there's a separate metadata manager under the hood, and that's way out of scope of the docs). At the runtime point of assignment, then the sanity check of type etc takes place, and not before.

So essentially you can assign what you like to a metadata instance, it doesn't care and it doesn't have to... it's only class container that **requires** to care. That's why I'm able to assign tuples to a metadata instance resulting from the `difference` operation :wink: 

Does that help de-weirded-out you?

> 

Reviewer 2: Thanks @bjlittle I will sleep better now! 😆

For example, given the following :class:`~iris.common.metadata.DimCoordMetadata` of the
``longitude`` coordinate,

.. doctest:: metadata-assign

    >>> longitude.metadata
    DimCoordMetadata(standard_name='longitude', long_name=None, var_name='longitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

We can assign to it directly using the :class:`~iris.common.metadata.DimCoordMetadata` of the ``latitude``
coordinate,

.. doctest:: metadata-assign

    >>> longitude.metadata = latitude.metadata
    >>> longitude.metadata
    DimCoordMetadata(standard_name='latitude', long_name=None, var_name='latitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Assign by iterable
""""""""""""""""""

It is also possible to assign to the ``metadata`` property of an Iris
`CF Conventions`_ container with an iterable containing the **correct
number** of associated member values, e.g.,

.. doctest:: metadata-assign

    >>> values = [getattr(latitude, member) for member in latitude.metadata._fields]
    >>> longitude.metadata = values
    >>> longitude.metadata

>     DimCoordMetadata(standard_name='latitude', long_name=None, var_name='latitude', units=Unit('degrees'), attributes={}, coord_system=GeogCS(6371229.0), climatological=False, circular=False)

Author: To be honest I think that's always (maybe) been there... anyways, there it is in all it's magnificent glory
